### PR TITLE
[EASI-3944] Notifications - On Discussion Replied Notification

### PR DESCRIPTION
**Pull request by OddTomBrooks**
# EASI-3944

## Changes and Description

* Created new notification meta structure
* Updated and generated GQL
* Updated plan discussion routing to gate sending reply emails based on preferences
* Updated plan discussion to send reply notifications based on preferences

## How to test this change

* Start the application
* Using Postman:
  * Create a default model plan
  * Create a default discussion
  * Create a default discussion reply
  * Query notifications
* Assert postman results are as expected
* Assert unit tests pass as expected

## PR Author Review Checklist

- [ ] Met the ticket's acceptance criteria, or will meet them in a subsequent PR.
- [ ] Added or updated tests for backend resolvers or other functions as needed.
- [ ] Added or updated client tests for new components, parent components, functions, or e2e tests as necessary.
- [ ] Updated the [Postman Collection](../MINT.postman_collection.json) if necessary.

## PR Reviewer Guidelines
- It's best to pull the branch locally and test it, rather than just looking at the code online!
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
- Always make comments, even if it's minor, or if you don't understand why something was done.

### CODE REVIEW

File: pkg/notifications/new_discussion_replied_meta.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+	"github.com/google/uuid"
+
+	"github.com/cmsgov/mint-app/pkg/models"
+	"github.com/cmsgov/mint-app/pkg/sqlutils"
+)
+
+// ActivityNewDiscussionRepliedCreate creates an activity for when a Discussion is replied to.
+func ActivityNewDiscussionRepliedCreate(ctx context.Context, np sqlutils.NamedPreparer, actorID uuid.UUID, discussionID uuid.UUID, replyID uuid.UUID, discussionContent models.TaggedHTML, getPreferencesFunc GetUserNotificationPreferencesFunc) (*models.Activity, error) {
+
+	activity := models.NewNewDiscussionRepliedActivity(actorID, discussionID, replyID, discussionContent.RawContent.String())
```
Comment (StevenWadeOddball) [inline pkg/notifications/new_discussion_replied_meta.go:15]: I think this should actually be the discussion reply content, not the discussion content.

File: pkg/notifications/new_discussion_replied_meta.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+	retActivity, actErr := activityCreate(ctx, np, activity)
+	if actErr != nil {
+		return nil, actErr
+	}
+
```
Comment (StevenWadeOddball) [inline pkg/notifications/new_discussion_replied_meta.go:21]: This method should also be responsible for creating the notification as well. I think it should go to the created user of the discussion.

File: pkg/notifications/new_discussion_replied_meta_test.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+import (
+	"context"
+
+	"github.com/google/uuid"
+
+	"github.com/cmsgov/mint-app/pkg/models"
+)
+
+func (suite *NotificationsSuite) TestActivityNewDiscussionRepliedCreate() {
```
Comment (StevenWadeOddball) [inline pkg/notifications/new_discussion_replied_meta_test.go:11]: This is a good start, you probably could do without any tags in this test, the other activity type really deals with the tagged notification. I'd expect assertions that
1. The user that made the discussion gets a notification for the reply, and then assert the content matches as expected.

File: pkg/models/new_discussion_replied_meta.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+import (
+	"database/sql/driver"
+	"encoding/json"
+	"errors"
+
+	"github.com/google/uuid"
+)
+
+// NewDiscussionRepliedActivityMeta represents the notification data that is relevant to a new Discussion Reply
+type NewDiscussionRepliedActivityMeta struct {
```
Comment (StevenWadeOddball) [inline pkg/models/new_discussion_replied_meta.go:12]: This looks pretty good! I'd do a final pass looking at [Figma](https://www.figma.com/file/8MfXzXzmKWv8vBjTasUIkD/MINT---Notifications?type=design&node-id=3472-13198&mode=design&t=q13knZsma5QK1gdV-0) and include any other fields that might be needed. At A glance, I think we might need model_plan_ID to be stored, as well as a resolver to modelPlan.

I added this onto Gary's [PR](https://github.com/CMSgov/mint-app/pull/988) for the reply notification type, I didn't realize it was needed, but it's important for rendering.

File: pkg/graph/resolvers/plan_discussion.go [hunk 5]
```diff
@@ -454,63 +454,73 @@ func CreateDiscussionReply(
...
-				modelPlan,
-				replyUser,
-			)
-			if errReplyEmail != nil {
-				logger.Error("error sending tagged in plan discussion reply emails to tagged users and teams",
-					zap.String("discussionID", discussion.ID.String()),
-					zap.Error(errReplyEmail))
-			}
+		// Early return if
+		if !pref.NewDiscussionReply.SendEmail() {
```
Comment (StevenWadeOddball) [inline pkg/graph/resolvers/plan_discussion.go:470]: This is going to make some issues, the tag emails have their own preferences, so this specific preference shouldn't block those emails. (The tag emails are based on the preference of each individual who is tagged, while is this should be based on the preference of the user who created the discussion.

File: pkg/graph/resolvers/plan_discussion.go [hunk 6]
```diff
@@ -454,63 +454,73 @@ func CreateDiscussionReply(
...
-
-			replyUser := principal.Account()
-			commonName := replyUser.CommonName
 
-			if err != nil {
-				logger.Error("error sending discussion reply emails. Unable to retrieve modelPlan",
-					zap.String("replyID", reply.ID.String()),
-					zap.Error(err))
-			}
+		pref, err := UserNotificationPreferencesGetByUserID(ctx, principal.Account().ID)
```
Comment (StevenWadeOddball) [inline pkg/graph/resolvers/plan_discussion.go:464]: I think this is getting preferences for the wrong user, it should be getting the preferences of the user who created the discussion, not the preferences of the user who made the reply. The discussion creator should get the notification

File: pkg/graph/resolvers/plan_discussion.go [hunk 7]
```diff
@@ -454,63 +454,73 @@ func CreateDiscussionReply(
...
+		if err != nil {
+			return nil, fmt.Errorf("unable to get user notification preference, Notification not created %w", err)
+		}
```
Comment (StevenWadeOddball) [inline pkg/graph/resolvers/plan_discussion.go:467]: Somewhere in this flow, the notification method needs to get called still.  `notifications.ActivityNewDiscussionRepliedCreate`

the actor should be the user who made the reply, but the notification should go to the user who created the discussion.

**Review by StevenWadeOddball — COMMENTED** [review]
Thanks @OddTomBrooks, this looks like you are on the right track!

I added some more notes as you continue to finalize the implementation.

The other thing not called out is just updating postman as well.

**Comment by StevenWadeOddball** [comment]
Review is in progress still. @OddTomBrooks, can you target the feature branch here please? This shouldn't merge to main until the frontend is built out.

File: pkg/graph/resolvers/plan_discussion.go [hunk 8]
```diff
@@ -454,13 +454,24 @@ func CreateDiscussionReply(
 		if err != nil {
 			return reply, err
 		}
+
 		// Create Activity and notifications in the DB
-		_, notificationErr := notifications.ActivityTaggedInDiscussionReplyCreate(ctx, tx, principal.Account().ID, discussion.ModelPlanID, discussion.ID, reply.ID, reply.Content, loaders.UserNotificationPreferencesGetByUserID)
+		_, notificationErr := notifications.ActivityTaggedInDiscussionReplyCreate(ctx, tx, discussion.CreatedBy, discussion.ModelPlanID, discussion.ID, reply.ID, reply.Content, loaders.UserNotificationPreferencesGetByUserID)
```
Comment (StevenWadeOddball) [inline pkg/graph/resolvers/plan_discussion.go:459]: This method call was updated incorrectly, `principal.Account().ID` is the actor, you switched it to `discussion.CreatedBy`. Can you please revert the change you made to this line?

File: pkg/graph/resolvers/activity.resolvers.go [hunk 9]
```diff
@@ -60,5 +84,16 @@ func (r *Resolver) TaggedInPlanDiscussionActivityMeta() generated.TaggedInPlanDi
...
 type taggedInDiscussionReplyActivityMetaResolver struct{ *Resolver }
 type taggedInPlanDiscussionActivityMetaResolver struct{ *Resolver }
+
+// !!! WARNING !!!
+// The code below was going to be deleted when updating resolvers. It has been copied here so you have
+// one last chance to move it out of harms way if you want. There are two reasons this happens:
+//   - When renaming or deleting a resolver the old code will be put in here. You can safely delete
+//     it when you're done.
+//   - You have helper methods in this file. Move them out to keep these resolver files clean.
+func (r *newDiscussionRepliedActivityMetaResolver) ModelPlanID(ctx context.Context, obj *models.NewDiscussionRepliedActivityMeta) (uuid.UUID, error) {
```
Comment (StevenWadeOddball) [inline pkg/graph/resolvers/activity.resolvers.go:97]: Can you delete this please?

File: pkg/notifications/notifications_test.go [hunk 10]
```diff
@@ -27,3 +29,14 @@ func TestNotificationsSuite(t *testing.T) {
 
 	suite.Run(t, rs)
 }
+
+func (suite *NotificationsSuite) deserializeActivityMetadata(testActivity *models.Activity) models.ActivityMetaData {
```
Comment (StevenWadeOddball) [inline pkg/notifications/notifications_test.go:33]: Optional: Could you add a comment to this test function? That would be helpful if the hope is to use this in other functions where the same logic is being tested.

**Review by StevenWadeOddball — CHANGES_REQUESTED** [review]
Thanks @OddTomBrooks, this is almost there!

I functionally tested the change with postman. I confirmed that a notification is created as expected, and that the email respects the user preference to only send an email when it is activated.

**Comment by OddTomBrooks** [comment]
> Thanks @OddTomBrooks, this is almost there!
> 
> I functionally tested the change with postman. I confirmed that a notification is created as expected, and that the email respects the user preference to only send an email when it is activated.

Thank you for the thorough review. All of your requests are in and ready for re-review, a request has also been sent.

**Review by StevenWadeOddball — APPROVED** [review]
Thanks for making the changes @OddTomBrooks, this looks good to go.

**Pull request merged**